98_spire_integration.md

```diff
@@ -0,0 +1,227 @@
+
+#### Story 1
+
+A developer will be able to develop a SPIRE agent and server plugin
+that communitcates with the Keylime agent and verifier to be able to
+independenly prove that the agent in question is on the same node as the
+SPIRE agent and also that the agent is passing it's attestation policies
+in Keylime.
+
+This integration will also pull in various properties of the Keylime setup
+(agent configuration, policy, etc) to use as selectors for SPIRE.
+
+### Notes/Constraints/Caveats (optional)
+
+None
+
+
+### Risks and Mitigations
+
+Care will need to be taken so that we don't leak any sensitive data in
+these APIs and that our verification/signing process is secure and leads
+to the guarantees we are making (that the SPIRE and Keylime agents are
+on the same node).
+
+The security of the information flows has been reviewed by several
+members of the Keylime development team as well as SPIRE participants. The
+implementation will need thorough review as well.
+
+
+## Design Details
+
+The following flow is anticipated for the full Keylime SPIRE plugins:
+
+```
+┌───────────────────────────────────────────────┐                    ┌───────────────┐
+│                                               │                    │               │
+│                     Node              #3      │                    │    SPIRE      │
+│           ┌───────────────────────────────────┼────────────────────►    SERVER     │
+│  ┌────────┴────┐                              │                    │               │
+│  │  SPIRE      │       #1                     │                    │               │
+│  │  Agent      ◄────────────────┐             │                    └─────┬─────────┘
+│  │             │                │             │                          │
+│  └─────────┬───┘                │             │                          │ #4
+│            │#2             ┌────▼─────┐       │                          │
+│         ┌──▼──────┐        │ Keylime  │       │                    ┌─────▼─────────┐
+│         │  TPM    │        │  Agent   │       │                    │               │
+│         │         │        │          │       │                    │    Keylime    │
+│         └─────────┘        └──────────┘       │                    │   Verifier    │
+│                                               │                    │               │
+└───────────────────────────────────────────────┘                    └───────────────┘
+```
+
+This flow has the following steps:
+
+1. SPIRE Agent queries node-local /info API on keylime agent to get information like the Keylime UUID
+2. SPIRE Agent creates a nonce that is sent to the TPM’s AK (keylime created) for signing
+3. SPIRE Agent sends the information to the Spire Server
+4. SPIRE Server queries Keylime Verifier about the agent. Does it exist? Is it passing attestation? If so, can you unencrypt (verify signature) of this nonce? If all are true, then SPIRE attestation passed and identity is issued.
```

Comment:
Is SPIRE attestation as 'periodic' as Keylime attestation?

Reply:
No, especially for node attestation, I believe it only happens once at SPIRE agent startup.

Reply:
Whatever SPIRE states about the node should only have a short-term validity period since Keylime attestation may detect that the nodes has gone out-of-policy shortly after ...